hoomd/mpcd/SlitPoreGeometryFiller.cc

```diff
@@ -0,0 +1,215 @@
+            if (N_box != 0)
+                {
+                h_boxes.data[m_num_boxes] = clampbox;
+                h_ranges.data[m_num_boxes] = make_uint2(m_N_fill, m_N_fill + N_box);
+                ++m_num_boxes;
+
+                m_N_fill += N_box;
+                }
+            }
+        }
+
+    // size is now updated, cache the cell dimensions used
+    m_needs_recompute = false;
+    m_last_cell_size = make_scalar2(cell_size, max_shift);
+    }
+
+/*!
+ * \param timestep Current timestep to draw particles
+ */
+void mpcd::SlitPoreGeometryFiller::drawParticles(unsigned int timestep)
+    {
+    // quit early if not filling to ensure we don't access any memory that hasn't been set
+    if (m_N_fill == 0) return;
+
+    ArrayHandle<Scalar4> h_pos(m_mpcd_pdata->getPositions(), access_location::host, access_mode::readwrite);
+    ArrayHandle<Scalar4> h_vel(m_mpcd_pdata->getVelocities(), access_location::host, access_mode::readwrite);
+    ArrayHandle<unsigned int> h_tag(m_mpcd_pdata->getTags(), access_location::host, access_mode::readwrite);
+    const Scalar vel_factor = fast::sqrt(m_T->getValue(timestep) / m_mpcd_pdata->getMass());
+
+    const BoxDim& box = m_pdata->getBox();
+    Scalar3 lo = box.getLo();
+    Scalar3 hi = box.getHi();
+
+    // boxes for filling
+    ArrayHandle<Scalar4> h_boxes(m_boxes, access_location::host, access_mode::read);
+    ArrayHandle<uint2> h_ranges(m_ranges, access_location::host, access_mode::read);
+    // set these counters so that they get filled on the first pass
+    int boxid = -1;
+    unsigned int boxlast = 0;
+
+    // index to start filling from
+    const unsigned int first_idx = m_mpcd_pdata->getN() + m_mpcd_pdata->getNVirtual() - m_N_fill;
+    for (unsigned int i=0; i < m_N_fill; ++i)
+        {
+        const unsigned int tag = m_first_tag + i;
+        hoomd::RandomGenerator rng(hoomd::RNGIdentifier::SlitPoreGeometryFiller, m_seed, tag, timestep);
+
+        // advanced past end of this box range, take the next
+        if (i >= boxlast)
+            {
+            ++boxid;
+            boxlast = h_ranges.data[boxid].y;
+            const Scalar4 fillbox = h_boxes.data[boxid];
+            lo.x = fillbox.x;
+            hi.x = fillbox.y;
+            lo.z = fillbox.z;
+            hi.z = fillbox.w;
+            }
+
+        const unsigned int pidx = first_idx + i;
+        h_pos.data[pidx] = make_scalar4(hoomd::UniformDistribution<Scalar>(lo.x,hi.x)(rng),
+                                        hoomd::UniformDistribution<Scalar>(lo.y,hi.y)(rng),
+                                        hoomd::UniformDistribution<Scalar>(lo.z,hi.z)(rng),
+                                        __int_as_scalar(m_type));
+
+        hoomd::NormalDistribution<Scalar> gen(vel_factor, 0.0);
+        Scalar3 vel;
+        gen(vel.x, vel.y, rng);
+        vel.z = gen(rng);
+        // TODO: should these be given zero net-momentum contribution (relative to the frame of reference?)
```

Comment:
Does this TODO item need to be resolved?

Reply:
There's also a similar comment in `SlitGeometryFiller`--I decided before that I don't think it's necessary to zero these (as in the Langevin thermostat, they are zero *on average*), but I left the note in case I came back to it later. I can remove if you prefer not to have an open TODO in the code.